sql/src/main/java/io/crate/executor/transport/task/elasticsearch/EsJobContextTask.java

```diff
@@ -0,0 +1,94 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.executor.transport.task.elasticsearch;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.crate.executor.JobTask;
+import io.crate.executor.TaskResult;
+import io.crate.jobs.ESJobContext;
+import io.crate.jobs.JobContextService;
+import io.crate.jobs.JobExecutionContext;
+import io.crate.operation.projectors.FlatProjectorChain;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.support.TransportAction;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.UUID;
+
+public class EsJobContextTask extends JobTask {
+
+    protected final List<SettableFuture<TaskResult>> results;
+    protected final int executionPhaseId;
+    private final JobContextService jobContextService;
+    protected JobExecutionContext context;
+
+    public EsJobContextTask(UUID jobId,
+                            int executionPhaseId,
+                            int numResults,
+                            JobContextService jobContextService) {
+        super(jobId);
+        this.executionPhaseId = executionPhaseId;
+        this.jobContextService = jobContextService;
+        results = new ArrayList<>(numResults);
+    }
+
+    protected void createContext(String operationName,
+                                 List<? extends ActionRequest> requests,
+                                 List<? extends ActionListener> listeners,
+                                 TransportAction transportAction,
+                                 @Nullable FlatProjectorChain projectorChain) {
+        ESJobContext esJobContext = new ESJobContext(operationName,
+                requests, listeners, results, transportAction, projectorChain);
+        JobExecutionContext.Builder contextBuilder = jobContextService.newBuilder(jobId());
+        contextBuilder.addSubContext(executionPhaseId, esJobContext);
+        context = jobContextService.createContext(contextBuilder);
+    }
+
+    @Override
+    final public void start() {
+        assert context != null : "Context must be created first";
+        try {
+            context.start();
+        } catch (Throwable throwable) {
+            for (SettableFuture<TaskResult> result : results) {
+                result.setException(throwable);
+            }
+        }
+    }
+
+    @Override
+    final public List<? extends ListenableFuture<TaskResult>> result() {
+        return results;
```

Comment:
maybe it would be possible to add an assertion here that the results aren't empty to ensure that subclasses added results to the list?